### PR TITLE
TF: tests for (de)serializable models with resized tokens

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Since I'm touching the part of our code that handles resizing of token embeddings, I decided to boost our test suite there.

This PR adds two tests regarding resizing token embeddings, on the TF side:
1. Tests that we can resize the embeddings of a model, save it, and then restore it (with the resized embeddings), while keeping the same outputs for a given input in the resized range;
2. Tests that passing inputs outside the vocabulary triggers an exception -- surprisingly, TF doesn't do this check on GPU, which means that a user can resize the embeddings incorrectly and run forward passes (inference or training) with incorrect numerical results, but no exceptions.

⚠️ The tests added above do not pass in all cases, and fixes will be arriving over subsequent PRs. To ensure it doesn't manifest in our push CI, the following touches were added -- alternative suggestions are welcome!
- Test 1. is failing for several models, so a `@slow` decorator was added to keep track of the failures while allowing the push CI to pass. It seemed more sensible to me than to add `skip` in all failing cases 🤔 
- Test 2. Fails for all models with embeddings except BART, on GPU. Our push CI doesn't use GPU, so it is not impacted.

### CODE REVIEW

File: src/transformers/modeling_tf_utils.py [hunk 1]
```diff
@@ -1889,6 +1890,40 @@ def _get_resized_lm_head_bias(self, old_lm_head_bias, new_num_tokens):
 
         return new_lm_head_bias
 
+    def _v2_get_resized_lm_head_bias(
```
Comment (gante) [inline src/transformers/modeling_tf_utils.py:1893]: The original `get_resized_lm_head_bias` actually adds weights to the model. Since those weights either have the same name (duplicated names -> save_weights fails) or a new name (i.e. can't map the new bias to the original bias at load time), model restoration with resized bias fails.

Also, the previous function was setting the bias with `trainable=True`, which is not the case for most models 😱 

The new function simply returns the new bias values, and delegates setting the bias to the model itself. For correct weight restoring capabilities, the bias weight has to be overwritten. Since different models have different bias layers, only the model itself can instantiate the layer -- see Bart's `set_bias()` changes.

File: src/transformers/models/bart/modeling_tf_bart.py [hunk 2]
```diff
@@ -1279,7 +1299,6 @@ def __init__(self, config, load_weight_prefix=None, *inputs, **kwargs):
         self.bias_layer = BiasLayer(
             name="final_logits_bias", shape=[1, config.vocab_size], initializer="zeros", trainable=False
         )
-        self.final_logits_bias = self.bias_layer.bias  # alias to keep the same interface with PT
```
Comment (gante) [inline src/transformers/models/bart/modeling_tf_bart.py:1282]: this alias was leaving traces of the original bias when it was overwritten, causing weight saving to fail (two variables with the same name)

File: tests/models/vit_mae/test_modeling_tf_vit_mae.py [hunk 3]
```diff
@@ -375,6 +375,7 @@ def test_keras_save_load(self):
 
     # overwrite from common since TFViTMAEForPretraining has random masking, we need to fix the noise
     # to generate masks during test
+    @slow
```
[the remark this answers is not in the document]
Comment (sgugger) [reply tests/models/vit_mae/test_modeling_tf_vit_mae.py:378]: Yes, this is fixed on main now.

**Review by sgugger — APPROVED** [review]
Thanks for the fixes!

File: src/transformers/modeling_tf_utils.py (hunk 1 above)
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:1893]: Wait, the bias on the LM head isn't trainable on most models? Is it initialized to match the frequency of tokens in the corpus and then frozen?

File: tests/test_modeling_tf_common.py [hunk 4]
```diff
@@ -1203,6 +1203,65 @@ def _get_word_embedding_weight(model, embedding_layer):
...
+                restored_model_outputs = model(**prepared_inputs)
+
+                # check that the output for the restored model is the same
+                self.assert_outputs_same(restored_model_outputs, outputs)
+
+    @unittest.skipIf(
+        not is_tf_available() or len(tf.config.list_physical_devices("GPU")) == 0,
+        reason="This test always passes on CPU.",
+    )
+    def test_embeddings_out_of_bounds_raise_exception(self):
```
Comment (Rocketknight1) [inline tests/test_modeling_tf_common.py:1247]: I still can't believe this isn't at least a flag that users could enable in TF

**Review by Rocketknight1 — APPROVED** [review]
Overall this looks good to me! I have a couple of comments but they're mostly about my own understanding rather than problems with the PR.

File: src/transformers/modeling_tf_utils.py (hunk 1 above)
Comment (gante) [reply src/transformers/modeling_tf_utils.py:1893]: Correct, depending on the model (e.g. [BERT's bias is trainable](https://github.com/huggingface/transformers/blob/70ba10e6d4684f56c50b740793498d4d2dd8a7bd/src/transformers/models/bert/modeling_tf_bert.py#L651)). When it is not trainable it can still be set, even if it stays as `0` the whole training procedure.

From the discussion on [slack](https://huggingface.slack.com/archives/C014N4749J9/p1661941064425229) regarding these untrainable bias: 
1. "the bias is saved as buffer [i.e. not trainable] because BART ties lm head with token embeddings, so the lm_head layer can’t have bias. Hence it’s added after applying  the lm_head."
2. "yeah and it's also explicitly a buffer to not compute gradients on it (...)"
3. "The bias is never updated at train time (...)"
